File: typescript/src/stream_redirect.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, statSync, symlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { configFromRawArgs } from "./stream_redirect.ts";

const dir = dirname(fileURLToPath(import.meta.url));

describe("stream redirect args", () => {
  it("parses space-separated and equals values", () => {
    const config = configFromRawArgs([
      "agent-cli",
      "--stdout-file",
      "/tmp/agent-cli.out",
      "--stderr-file=/tmp/agent-cli.err",
      "ping",
    ]);
    assert.deepEqual(config, {
      stdoutFile: "/tmp/agent-cli.out",
      stderrFile: "/tmp/agent-cli.err",
    });
  });

  it("returns undefined when disabled", () => {
    assert.equal(configFromRawArgs(["agent-cli", "ping"]), undefined);
  });

  it("rejects missing values", () => {
    assert.throws(() => configFromRawArgs(["agent-cli", "--stderr-file", "--help"]));
  });

  it("redirects stdout and native stderr in a child process", () => {
    // fd-level stdout/stderr redirection is a POSIX capability; Node on Windows
    // cannot reassign the process stdio fds the way this relies on.
    if (process.platform === "win32") return;
    const tempDir = mkdtempSync(join(tmpdir(), "afdata-stream-redirect-"));
    const stdoutPath = join(tempDir, "stdout.log");
    const stderrPath = join(tempDir, "stderr.log");
    const childPath = join(tempDir, "redirect_child.ts");
    const tsxBin = join(dir, "..", "node_modules", ".bin", process.platform === "win32" ? "tsx.cmd" : "tsx");

    if (!existsSync(tsxBin)) {
      throw new Error(`tsx executable not found at ${tsxBin}`);
    }

    writeFileSync(
      childPath,
      [
        `import { installStreamRedirectFromRawArgs } from ${JSON.stringify(pathToFileURL(join(dir, "stream_redirect.ts")).href)};`,
        "installStreamRedirectFromRawArgs(process.argv.slice(2));",
        'process.stdout.write("stdout bytes\\n");',
        'throw new Error("stderr bytes");',
        "",
      ].join("\n"),
    );
    writeFileSync(stdoutPath, "existing stdout\n");

    try {
      execFileSync(tsxBin, [childPath, "--stdout-file", stdoutPath, "--stderr-file", stderrPath], {
        encoding: "utf-8",
        stdio: "pipe",
      });
      assert.fail("child process should fail after throwing");
    } catch (error) {
      assert.ok(error instanceof Error && "stdout" in error && "stderr" in error);
      const output = error as Error & { stdout: string; stderr: string };
      assert.equal(output.stdout, "");
      assert.equal(output.stderr, "");
    }

    assert.equal(readFileSync(stdoutPath, "utf-8"), "existing stdout\nstdout bytes\n");
    assert.match(readFileSync(stderrPath, "utf-8"), /Error: stderr bytes/);
    if (process.platform !== "win32") {
      assert.equal(statSync(stderrPath).mode & 0o777, 0o600);
    }
  });

  it("rejects symbolic link targets", () => {
    if (process.platform === "win32") return;
    const tempDir = mkdtempSync(join(tmpdir(), "afdata-stream-redirect-symlink-"));
    const realPath = join(tempDir, "real.log");
    const symlinkPath = join(tempDir, "stdout.log");
    const childPath = join(tempDir, "redirect_symlink_child.ts");
    const tsxBin = join(dir, "..", "node_modules", ".bin", "tsx");

    writeFileSync(realPath, "");
    symlinkSync(realPath, symlinkPath);
    writeFileSync(
      childPath,
      [
        `import { installStreamRedirectFromRawArgs } from ${JSON.stringify(pathToFileURL(join(dir, "stream_redirect.ts")).href)};`,
        "installStreamRedirectFromRawArgs(process.argv.slice(2));",
        'process.stdout.write("should not redirect\\n");',
        "",
      ].join("\n"),
    );

    assert.throws(
      () => execFileSync(tsxBin, [childPath, "--stdout-file", symlinkPath], { encoding: "utf-8", stdio: "pipe" }),
      /symbolic link/,
    );
    assert.equal(readFileSync(realPath, "utf-8"), "");
  });
});

File: go/streamredirect/streamredirect_test.go
package streamredirect

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func TestParseStreamRedirectArgs(t *testing.T) {
	cfg, err := ParseStreamRedirectArgs([]string{
		"agent-cli",
		"--stdout-file",
		"/tmp/agent-cli.out",
		"--stderr-file=/tmp/agent-cli.err",
		"ping",
	})
	if err != nil {
		t.Fatalf("ParseStreamRedirectArgs returned error: %v", err)
	}
	if cfg == nil {
		t.Fatal("expected stream redirection config")
	}
	if cfg.StdoutFile != "/tmp/agent-cli.out" {
		t.Fatalf("StdoutFile = %q", cfg.StdoutFile)
	}
	if cfg.StderrFile != "/tmp/agent-cli.err" {
		t.Fatalf("StderrFile = %q", cfg.StderrFile)
	}
}

func TestParseStreamRedirectArgsDisabled(t *testing.T) {
	cfg, err := ParseStreamRedirectArgs([]string{"agent-cli", "ping"})
	if err != nil {
		t.Fatalf("ParseStreamRedirectArgs returned error: %v", err)
	}
	if cfg != nil {
		t.Fatalf("expected nil config, got %#v", cfg)
	}
}

func TestParseStreamRedirectArgsMissingValue(t *testing.T) {
	if _, err := ParseStreamRedirectArgs([]string{"agent-cli", "--stderr-file", "--help"}); err == nil {
		t.Fatal("expected missing value error")
	}
}

func TestInstallStreamRedirectUnsupportedOnWindows(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("windows-only unsupported-platform assertion")
	}
	installed, err := InstallStreamRedirect(StreamRedirectConfig{StdoutFile: "stdout.log"})
	if installed != nil {
		_ = installed.Close()
	}
	if err == nil {
		t.Fatal("expected unsupported stream redirection error")
	}
	if !strings.Contains(err.Error(), "only supported on Unix") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestInstallStreamRedirectRedirectsOutput(t *testing.T) {
	dir := t.TempDir()
	stdoutPath := filepath.Join(dir, "stdout.log")
	stderrPath := filepath.Join(dir, "stderr.log")
	if err := os.WriteFile(stdoutPath, []byte("existing stdout\n"), 0o644); err != nil {
		t.Fatalf("prewrite stdout file: %v", err)
	}

	installed, err := InstallStreamRedirect(StreamRedirectConfig{
		StdoutFile: stdoutPath,
		StderrFile: stderrPath,
	})
	if err != nil {
		if strings.Contains(err.Error(), "only supported on Unix") {
			t.Skip(err)
		}
		t.Fatalf("InstallStreamRedirect returned error: %v", err)
	}
	closed := false
	defer func() {
		if !closed {
			_ = installed.Close()
		}
	}()

	if _, err := os.Stdout.WriteString("stdout bytes\n"); err != nil {
		t.Fatalf("write redirected stdout: %v", err)
	}
	if _, err := os.Stderr.WriteString("stderr bytes\n"); err != nil {
		t.Fatalf("write redirected stderr: %v", err)
	}
	if err := installed.Close(); err != nil {
		t.Fatalf("Close returned error: %v", err)
	}
	closed = true

	stdoutData, err := os.ReadFile(stdoutPath)
	if err != nil {
		t.Fatalf("read stdout file: %v", err)
	}
	if string(stdoutData) != "existing stdout\nstdout bytes\n" {
		t.Fatalf("stdout file = %q", stdoutData)
	}
	stderrData, err := os.ReadFile(stderrPath)
	if err != nil {
		t.Fatalf("read stderr file: %v", err)
	}
	if string(stderrData) != "stderr bytes\n" {
		t.Fatalf("stderr file = %q", stderrData)
	}
	stderrInfo, err := os.Stat(stderrPath)
	if err != nil {
		t.Fatalf("stat stderr file: %v", err)
	}
	if stderrInfo.Mode().Perm() != 0o600 {
		t.Fatalf("stderr file mode = %o, want 600", stderrInfo.Mode().Perm())
	}
}

func TestInstallStreamRedirectRejectsSymlink(t *testing.T) {
	dir := t.TempDir()
	realPath := filepath.Join(dir, "real.log")
	linkPath := filepath.Join(dir, "stdout.log")
	if err := os.WriteFile(realPath, []byte{}, 0o600); err != nil {
		t.Fatalf("create real file: %v", err)
	}
	if err := os.Symlink(realPath, linkPath); err != nil {
		t.Skipf("symlink unsupported: %v", err)
	}

	installed, err := InstallStreamRedirect(StreamRedirectConfig{StdoutFile: linkPath})
	if installed != nil {
		_ = installed.Close()
	}
	if err == nil {
		t.Fatal("expected symlink target error")
	}
	if strings.Contains(err.Error(), "only supported on Unix") {
		t.Skip(err)
	}
	if !strings.Contains(err.Error(), "symbolic link") && !strings.Contains(err.Error(), "too many") {
		t.Fatalf("unexpected error: %v", err)
	}
}
